testsuite/do_tests.sh

```diff
@@ -0,0 +1,684 @@
+            PYNEST_TEST_PASSED=$(($PYNEST_TEST_TOTAL - $PYNEST_TEST_SKIPPED - $PYNEST_TEST_FAILED))
+
+        else
+
+            echo
+            echo "  Nosetests unavailable. Using fallback test harness."
+            echo "  Fewer tests will be excuted."
+            echo
+
+            "${PYTHON}" "${PYTHON_HARNESS}" >> "${TEST_LOGFILE}"
+
+            PYNEST_TEST_TOTAL="$(  cat pynest_test_numbers.log | cut -d' ' -f1 )"
+            PYNEST_TEST_PASSED="$( cat pynest_test_numbers.log | cut -d' ' -f2 )"
+            PYNEST_TEST_SKIPPED="$( cat pynest_test_numbers.log | cut -d' ' -f3 )"
+            PYNEST_TEST_FAILED="$( cat pynest_test_numbers.log | cut -d' ' -f4 )"
+
+            rm -f "pynest_test_numbers.log"
+
+        fi
+
+        echo
+        echo "  PyNEST tests: ${PYNEST_TEST_TOTAL}"
+        echo "     Passed: ${PYNEST_TEST_PASSED}"
+        echo "     Skipped: ${PYNEST_TEST_SKIPPED}"
+        echo "     Failed: ${PYNEST_TEST_FAILED}"
+
+        PYNEST_TEST_SKIPPED_TEXT="(${PYNEST_TEST_SKIPPED} PyNEST)"
+        PYNEST_TEST_FAILED_TEXT="(${PYNEST_TEST_FAILED} PyNEST)"
+
+    else
+	echo
+	echo "Phase 7: PyNEST tests"
+	echo "---------------------"
+	echo "  Not running PyNEST tests because NEST was compiled without support"
+	echo "  for Python. See the file README.md for details."
+    fi
+}
+
+phase_eight() {
+    if command -v ${TEST_BASEDIR}/testsuite/cpptests/run_all_cpptests > /dev/null 2>&1; then
+        echo
+        echo "Phase 8: C++ tests (experimental)"
+        echo "---------------------------------"
+
+        CPP_TEST_OUTPUT="$(${TEST_BASEDIR}/testsuite/cpptests/run_all_cpptests 2>&1)"
+        # TODO:
+        # We should check the return code ($?) of run_all_cpptests here to see
+        # if a fatal crash occured. We cannot simply test $? != 0, since
+        # run_all_cpptests will return a non-zero code if tests fail.
+
+        # TODO:
+        # The regex for the total number seems fine according to
+        # https://www.boost.org/doc/libs/1_65_0/libs/test/doc/html/boost_test/test_output/log_formats/log_human_readable_format.html
+        # but does the check for failures work as it should?
+        # At some point, we should also count skipped tests.
+        CPP_TEST_TOTAL=$(echo "$CPP_TEST_OUTPUT" | sed -${EXTENDED_REGEX_PARAM} -n 's/.*Running ([0-9]+).*/\1/p')
+        CPP_TEST_FAILED=$(echo "$CPP_TEST_OUTPUT" | sed -${EXTENDED_REGEX_PARAM} -n 's/.*([0-9]+) failure.*/\1/p')
+
+        # replace empty strings by zero so arithmetic expressions below work
+        CPP_TEST_TOTAL=${CPP_TEST_TOTAL:-0}
+        CPP_TEST_FAILED=${CPP_TEST_FAILED:-0}
+        CPP_TEST_PASSED=$(( $CPP_TEST_TOTAL - $CPP_TEST_FAILED ))
+        CPP_TEST_SKIPPED=0
+
+        echo "  C++ tests : ${CPP_TEST_TOTAL}"
+        echo "     Passed : ${CPP_TEST_PASSED}"
+        echo "     Failed : ${CPP_TEST_FAILED}"
+    else
+	echo "  Not running C++ tests because NEST was compiled without Boost."
+    fi
+}
+
+phase_one
```

Comment:
Should we add to the name what the tests are doing? E.g. `phase_one_selftests`?

Reply:
I think that we should actually have some kind of associative array mapping phase names to phase functions and then automatic command line arguments along the lines of `--list-phases`and `--skip-phases`. Ideally the phase functions would be in the corresponding directories, accompanied by a metadata file. I would, however, find this to be out of the scope of this PR.

@terhorstd: what is your take on this?

Reply:
@jougs I like your idea! It would be useful to be able to selectively run only certain groups of tests.

Reply:
@heplesser: thanks for liking the idea. Should this also go into this PR or should it be its own?

Reply:
@jougs There is probably still enough to do in this PR, so I wouldn't want to enlarge. Create a separate one for the named and flexible phases.